[CTF] Fix out-of-bounds memory access in EncodedBlocks

MichaelLettrich opened the pull request:
Introduce padding bits to prevent out-of-bounds memory access in Encoded
blocks when writing in-compressible Symbols or data that is not passed
inside the entropy coder.
Additional cleanups.

MichaelLettrich commented:
@davidrohr this should fix the errors detected by the memory sanitizer. Can you have a look and see if it really is the case?

davidrohr commented:
I don't have the setup anymore to build with the sanitizer enabled. I could recheck, but can't we simply see from the code that what was fixed by Ruben is also fixed here? There was a lot of renaming, so without looking in detail it is not fully clear to me what is the actual fix.

MichaelLettrich commented on DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/EncodedBlocks.h:
```diff
@@ -808,59 +815,79 @@ void EncodedBlocks<H, N, W>::encode(const S_IT srcBegin,     // iterator begin o
-    *meta = Metadata{messageLength, literals.size(), sizeof(uint64_t), sizeof(stream_t), static_cast<uint8_t>(encoder->getSymbolTablePrecision()), opt,
-                     encoder->getMinSymbol(), encoder->getMaxSymbol(), dictSize, dataSize, literalSize};
+    // store incompressible symbols if any
+    const size_t nLiteralSymbols = [&]() {
+      const size_t nSymbols = literals.size();
+      if (!literals.empty()) {
+        // introduce padding in case literals don't align;
+        const size_t nPaddingElements = (nSymbols * sizeof(input_t)) % sizeof(storageBuffer_t);
+        literals.insert(literals.end(), nPaddingElements, {});
+
+        const size_t literalStorageSize = (literals.size() * sizeof(input_t)) / sizeof(storageBuffer_t) + (sizeof(input_t) < sizeof(storageBuffer_t));
+        expandStorage(literalStorageSize);
+        thisBlock->storeLiterals(literalStorageSize, reinterpret_cast<const storageBuffer_t*>(literals.data()));
+      }
+      return nSymbols;
+    }();
```
@davidrohr @shahor02  this introduces padding to the literal symbols

MichaelLettrich commented on DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/EncodedBlocks.h:
```diff
@@ -808,59 +815,79 @@ void EncodedBlocks<H, N, W>::encode(const S_IT srcBegin,     // iterator begin o
+    const int dataSize = messageSizeBytes / sizeof(storageBuffer_t) + (sizeof(input_t) < sizeof(storageBuffer_t));
     // no dictionary needed
     expandStorage(dataSize);
-    *meta = Metadata{messageLength, 0, sizeof(uint64_t), sizeof(stream_t), probabilityBits, opt, 0, 0, 0, dataSize, 0};
-    //FIXME: no we don't need an intermediate vector.
+    const size_t nPaddingElements = messageSizeBytes % sizeof(storageBuffer_t);
+    //FIXME(milettri): we should be able to do without an intermediate vector;
     // provided iterator is not necessarily pointer, need to use intermediate vector!!!
-    std::vector<STYP> vtmp(srcBegin, srcEnd);
-    bl->storeData(meta->nDataWords, reinterpret_cast<const W*>(vtmp.data()));
+    std::vector<input_t> vtmp(srcBegin, srcEnd);
+    vtmp.insert(vtmp.end(), nPaddingElements, {});
+    thisBlock->storeData(thisMetadata->nDataWords, reinterpret_cast<const W*>(vtmp.data()));
+
+    *thisMetadata = Metadata{messageLength, 0, sizeof(ransState_t), sizeof(storageBuffer_t), symbolTablePrecision, opt, 0, 0, 0, dataSize, 0};
```
@davidrohr @shahor02 this introduces padding to uncompressed messages.

MichaelLettrich commented:
@davidrohr  can you tell me how to run the sanitizer myself ? Did you use clang for that or did you use the GCC sanitizer?

shahor02 commented:
Hi @MichaelLettrich 
I am on the phone now, will check bit later. But you don't need sanitizer to validated this, valgrind also shows the problem

davidrohr commented:
Then valgrind is actually the easier thing to do.

In order to run the sanitizer:
- Add build flags `-O0 -fsanitize=address`.
- Building the OpenCL part and the analysis will fail due to sanitizer errors. You can add `| true` to the CMake add_custom_command rules for those. This will basically not create those files, but they are not needed for the TPC entropy encoding.
- The sanitizer library will not correctly load for the forked processes for some reason, so you need to `LD_PRELOAD=[GCC_LIB_PATH]/libasan.so`
- Then there are some ROOT things causing sanitizer warnings, it should print out some env settings you have to apply.
- Afterwards just run the workflow normally.

shahor02 commented on DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/EncodedBlocks.h:
```diff
@@ -808,59 +815,79 @@ void EncodedBlocks<H, N, W>::encode(const S_IT srcBegin,     // iterator begin o
+        const size_t nPaddingElements = (nSymbols * sizeof(input_t)) % sizeof(storageBuffer_t);
+        literals.insert(literals.end(), nPaddingElements, {});
+
+        const size_t literalStorageSize = (literals.size() * sizeof(input_t)) / sizeof(storageBuffer_t) + (sizeof(input_t) < sizeof(storageBuffer_t));
```
Same comment as before: here for `sizeof(input_t) < sizeof(storageBuffer_t)` you are always adding 1, while it should be added only if `(literals.size() * sizeof(input_t)) % sizeof(storageBuffer_t) != 0`.
See how I did this in https://github.com/AliceO2Group/AliceO2/blob/1e054d67466173870d49c8726f955f90de4f4f97/DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/EncodedBlocks.h#L848-L854

shahor02 commented on DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/EncodedBlocks.h:
```diff
@@ -808,59 +815,79 @@ void EncodedBlocks<H, N, W>::encode(const S_IT srcBegin,     // iterator begin o
-    const size_t szb = messageLength * sizeof(STYP);
-    const int dataSize = szb / sizeof(stream_t) + (sizeof(STYP) < sizeof(stream_t));
+    const size_t messageSizeBytes = messageLength * sizeof(input_t);
+    const int dataSize = messageSizeBytes / sizeof(storageBuffer_t) + (sizeof(input_t) < sizeof(storageBuffer_t));
```
same comment here: you add unneeded slot even if when there is no alignment problem.

shahor02 commented on DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/EncodedBlocks.h:
```diff
@@ -808,59 +815,79 @@ void EncodedBlocks<H, N, W>::encode(const S_IT srcBegin,     // iterator begin o
-    std::vector<STYP> vtmp(srcBegin, srcEnd);
-    bl->storeData(meta->nDataWords, reinterpret_cast<const W*>(vtmp.data()));
+    std::vector<input_t> vtmp(srcBegin, srcEnd);
+    vtmp.insert(vtmp.end(), nPaddingElements, {});
```
Since vtmp is a local vector with size given in constructor, `insert` will certainly lead to reallocation.
That's why to avoid it I calculate the padded size in the beginning, create a vector of the final size, then 
copy the data, the padding by 0s is there by construction, see https://github.com/AliceO2Group/AliceO2/blob/1e054d67466173870d49c8726f955f90de4f4f97/DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/EncodedBlocks.h#L861-L869

shahor02 reviewed (commented):
Hi @MichaelLettrich 
I think the padding size calculation should be changed, see inline comments.

To validate with valgring, you can run it with 
````
... | o2-ctf-writer-workflow  --ctf-writer '--child-driver "valgrind "'  $ARGS_ALL  --onlyDet $CTF_DETECTORS 
````
W/o correct fix it will show `invalid reads`

shahor02 commented on DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/EncodedBlocks.h:
```diff
@@ -70,6 +70,23 @@ inline T* relocatePointer(const char* oldBase, char* newBase, const T* ptr)
+inline size_t calculateNDestTElements(size_t sourceElems)
+{
+  const size_t sizeOfSourceArray = sourceElems * sizeof(source_T);
+  return sourceElems / sizeof(dest_T) + (sourceElems % sizeof(dest_T) != 0);
```
there should be `sizeOfSourceArray` here instead of `sourceElems`

MichaelLettrich replied:
Indeed. Stupid mistake thank you for spotting it!

shahor02 commented on DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/EncodedBlocks.h:
```diff
@@ -874,22 +891,20 @@ void EncodedBlocks<H, N, W>::encode(const input_IT srcBegin,      // iterator be
     std::vector<input_t> vtmp(srcBegin, srcEnd);
+    // introduce padding in case literals don't align;
+    const size_t nPaddingElements = calculatePadding<input_t, storageBuffer_t>(messageLength);
     vtmp.insert(vtmp.end(), nPaddingElements, {});
```
you still have 1 guaranteed re-allocation of vtmp here, why not booking it from very beginning with the right size?

MichaelLettrich replied:
see comment below.

shahor02 replied:
which comment? This `//FIXME(milettri): we should be able to do without an intermediate vector`? 
Are you going to change the storeData interface?

MichaelLettrich committed:
[CTF] Fix out-of-bounds access in EncodedBlocks

Add Padding to buffers to prevent out-of-bounds memory access in
EncodedBlocks when writing incompressible symbols or data that
is not handled by the entropy coder at all.

Additional cleanups.

MichaelLettrich commented:
@shahor02  thank you very much for the review especially spotting the stupid resizing mistake.

The vector sizes are interesting. What I have learned in my uni classes on algorithms and data structures was that a dynamic array like `std::vector` has both, a `size` and a `capacity`. `size` is the actual size of data it holds where as `capacity` refers to the size of the underlying allocated memory. To have low amortized cost of reallocation, insertion doubles the size of the container (see e.g. https://mortoray.com/2019/03/26/the-amortized-cost-of-vector-insert-is-3)
The standard itself doesn't state how the `capacity` is exactly affected by the constructor or  `resize`, `reserve` and `insert` apart from the obvious fact that `capacity >= size` after each operation.

Investigating things in compiler explorer yields surprisingly unsurprising results: https://godbolt.org/z/ccqqG5fP8
* `push_back` and `insert` behave as I would have expected from a dynamic array, doubling `capacity` upon resize.
* The constructor, `reserve` and `resize` behave as they intuitively would and really just allocate the exact amount of memory that is requested. I was expecting the same behavior as for `push_back` especially for the constructor. This btw seems to apply for GCC, clang, Intel and even MSVC. What this boils down to is that in case of the `literals` vector, which is created via `push_back`, `insert` will be most likely equivalent to `resize` whereas with the temporary vector it definitely will not. 

For the sake of readability I would have preferred the `insert` option as it seemed clearer to me that the new size will be larger than the old one.

This, at least to me was an interesting experiment and a really good to know fact about our most used container. I made the according changes to prevent resizing. Thank you !

shahor02 commented:
@MichaelLettrich yes, constructor always books exact size requested, only dynamically expansion methods preallocate (by doubling the capacity) the memory if new size would exceed the capacity.
